components/swim/src/server/mod.rs

```diff
@@ -0,0 +1,280 @@
+// Copyright (c) 2016 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! The SWIM server.
+//!
+//! Creates `Server` structs, that hold everything we need to run the SWIM protocol. Winds up with
+//! 3 separate threads - inbound (incoming connections), outbound (the Probe protocl), and expire
+//! (turning Suspect members into Confirmed members).
+
+pub mod expire;
+pub mod inbound;
+pub mod outbound;
+
+use std::collections::HashSet;
+use std::clone::Clone;
+use std::fmt;
+use std::net::{ToSocketAddrs, UdpSocket, SocketAddr};
+use std::sync::{Arc, RwLock};
+use std::sync::atomic::{AtomicBool, AtomicIsize, Ordering};
+use std::sync::mpsc::channel;
+use std::time::Duration;
+use std::thread;
+
+use error::{Result, Error};
+use member::{Member, Health, MemberList};
+use trace::Trace;
+use rumor::{RumorList, RumorKey};
+
+/// The server struct. Is thread-safe.
+#[derive(Debug)]
+pub struct Server {
+    pub name: Arc<RwLock<String>>,
+    pub socket: UdpSocket,
+    pub member: Arc<RwLock<Member>>,
+    pub member_list: MemberList,
+    pub rumor_list: RumorList,
+    // These are all here for testing support
+    pub pause: Arc<AtomicBool>,
+    pub trace: Arc<RwLock<Trace>>,
+    pub rounds: Arc<AtomicIsize>,
+    pub blacklist: Arc<RwLock<HashSet<SocketAddr>>>,
+}
+
+impl Clone for Server {
+    fn clone(&self) -> Server {
+        let socket = match self.socket.try_clone() {
+            Ok(socket) => socket,
+            Err(e) => {
+                error!("Failed to clone socket; trying again: {:?}", e);
+                return self.clone();
```

Comment:
Yep. One of the things I'm worried about here is that very little of this code should be able to fail. We really should only be panicing if we have locks we can't fix.